File: backend/app/schemas/course.py
"""User-facing course input and course record."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

TemplateKind = Literal["technical", "non_technical"]

TEMPLATE_IDS: dict[str, str] = {
    "technical": "technical_v1",
    "non_technical": "non_technical_v1",
}


class TocItem(BaseModel):
    """One chapter of the table of contents."""

    model_config = ConfigDict(extra="forbid")

    title: str = Field(min_length=1, max_length=300)
    sections: list[str] = Field(default_factory=list)
    notes: str = ""

    @field_validator("sections", mode="before")
    @classmethod
    def _coerce_sections(cls, value: object) -> object:
        if value is None:
            return []
        if isinstance(value, str):
            return [value]
        return value


class CourseInput(BaseModel):
    """The six MVP inputs."""

    model_config = ConfigDict(extra="forbid")

    course_title: str = Field(min_length=1, max_length=300)
    toc: list[TocItem] = Field(min_length=1)
    target_audience: str = Field(min_length=1, max_length=2000)
    dos: list[str] = Field(default_factory=list)
    donts: list[str] = Field(default_factory=list)
    template: TemplateKind
    language: str = "en"
    tone: str = ""

    @field_validator("toc", mode="before")
    @classmethod
    def _coerce_toc(cls, value: object) -> object:
        """Accept a plain list of chapter titles for convenience."""
        if isinstance(value, list):
            return [{"title": item} if isinstance(item, str) else item for item in value]
        return value

    @property
    def template_id(self) -> str:
        return TEMPLATE_IDS[self.template]

    def chapter_titles(self) -> list[str]:
        return [item.title for item in self.toc]


CourseStatus = Literal[
    "created",
    "planned",
    "researching",
    "writing",
    "reviewing",
    "assembling",
    "illustrating",
    "ready",
    "failed",
]


class ChapterProgress(BaseModel):
    model_config = ConfigDict(extra="allow")

    chapter_id: str
    title: str
    researched: bool = False
    written: bool = False
    reviewed: bool = False
    review_score: float | None = None
    error: str | None = None


class RunInfo(BaseModel):
    """Live state of the current or last generation run.

    Persisted alongside the course so the progress screen can show a real ETA and
    an interrupted run can be resumed.
    """

    model_config = ConfigDict(extra="allow")

    job_id: str = ""
    state: Literal["queued", "running", "done", "failed", "cancelled"] = "queued"
    mode: Literal["background", "sync"] = "background"
    started_at: str = ""
    updated_at: str = ""
    finished_at: str = ""
    chapters_total: int = 0
    chapters_done: int = 0
    writing_mode: str = "parallel"
    eta_seconds: float | None = None
    error: str | None = None
    timings: dict[str, Any] = Field(default_factory=dict)


class CourseRecord(BaseModel):
    """Everything we persist about a course besides blueprint/research/chapters."""

    model_config = ConfigDict(extra="allow")

    course_id: str
    document_id: str
    run: RunInfo | None = None
    status: CourseStatus = "created"
    input: CourseInput
    template_id: str
    created_at: str
    updated_at: str
    chapters: list[ChapterProgress] = Field(default_factory=list)
    has_blueprint: bool = False
    has_document: bool = False
    pdf_path: str | None = None
    last_error: str | None = None
    warnings: list[str] = Field(default_factory=list)


# --- API payloads -----------------------------------------------------------


class CreateCourseRequest(CourseInput):
    """Course creation runs the planner unless `run_planner` is false."""

    run_planner: bool = True


class ImproveTocRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    course_title: str = Field(min_length=1)
    toc: list[TocItem] = Field(default_factory=list)
    audience: str = ""
    template: TemplateKind = "technical"
    dos: list[str] = Field(default_factory=list)
    donts: list[str] = Field(default_factory=list)

    @field_validator("toc", mode="before")
    @classmethod
    def _coerce_toc(cls, value: object) -> object:
        if isinstance(value, list):
            return [{"title": item} if isinstance(item, str) else item for item in value]
        return value


class TocChange(BaseModel):
    model_config = ConfigDict(extra="allow")

    action: str  # add | remove | rename | reorder | split | merge
    target: str = ""
    proposed: str = ""
    reason: str = ""


class ImproveTocResponse(BaseModel):
    model_config = ConfigDict(extra="allow")

    suggested_toc: list[TocItem] = Field(default_factory=list)
    changes: list[TocChange] = Field(default_factory=list)
    reasoning: str = ""
    missing_concepts: list[str] = Field(default_factory=list)
    duplicate_topics: list[str] = Field(default_factory=list)


class GenerateRequest(BaseModel):
    """Drive a full or partial generation run."""

    model_config = ConfigDict(extra="forbid")

    chapter_ids: list[str] | None = None
    force: bool = False
    skip_research: bool = False
    generate_images: bool | None = None
    build_document: bool = True
    # background returns immediately with a job id and the caller polls
    # GET /api/courses/{id}; sync blocks until the run finishes.
    mode: Literal["background", "sync"] = "background"
    # Skip chapters that already have artifacts on disk (used when resuming an
    # interrupted run).
    resume: bool = False
    # Deep Research costs minutes per chapter, so it is opt-in per chapter rather
    # than a whole-course mode.
    deep_research_chapter_ids: list[str] | None = None


class GenerateResponse(BaseModel):
    model_config = ConfigDict(extra="allow")

    course_id: str
    document_id: str
    status: CourseStatus
    chapters_generated: list[str] = Field(default_factory=list)
    chapters_failed: list[str] = Field(default_factory=list)
    images_generated: int = 0
    pages: int = 0
    warnings: list[str] = Field(default_factory=list)
    duration_seconds: float = 0.0
    mode: Literal["background", "sync"] = "sync"
    job_id: str = ""
    accepted: bool = False  # true when the run was started in the background
    timings: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/lib/types/course.ts
/**
 * Mirrors backend/app/schemas/course.py and blueprint.py.
 * Keep these in sync with the Pydantic models.
 */

export type TemplateKind = "technical" | "non_technical";

export const TEMPLATE_IDS: Record<TemplateKind, string> = {
  technical: "technical_v1",
  non_technical: "non_technical_v1",
};

export interface TocItem {
  title: string;
  sections: string[];
  notes: string;
}

export interface CourseInput {
  course_title: string;
  toc: TocItem[];
  target_audience: string;
  dos: string[];
  donts: string[];
  template: TemplateKind;
  language: string;
  tone: string;
}

export interface CreateCourseRequest extends CourseInput {
  run_planner: boolean;
}

export type CourseStatus =
  | "created"
  | "planned"
  | "researching"
  | "writing"
  | "reviewing"
  | "assembling"
  | "illustrating"
  | "ready"
  | "failed";

export interface ChapterProgress {
  chapter_id: string;
  title: string;
  researched: boolean;
  written: boolean;
  reviewed: boolean;
  review_score: number | null;
  error: string | null;
}

export type RunState = "queued" | "running" | "done" | "failed" | "cancelled";

/** Live state of the current or last generation run, persisted by the backend. */
export interface RunInfo {
  job_id: string;
  state: RunState;
  mode: "background" | "sync";
  started_at: string;
  updated_at: string;
  finished_at: string;
  chapters_total: number;
  chapters_done: number;
  writing_mode: string;
  eta_seconds: number | null;
  error: string | null;
  timings: Record<string, unknown>;
}

export interface CourseRecord {
  course_id: string;
  document_id: string;
  run?: RunInfo | null;
  status: CourseStatus;
  input: CourseInput;
  template_id: string;
  created_at: string;
  updated_at: string;
  chapters: ChapterProgress[];
  has_blueprint: boolean;
  has_document: boolean;
  pdf_path: string | null;
  last_error: string | null;
  warnings: string[];
}

// --- TOC improvement --------------------------------------------------------

export interface ImproveTocRequest {
  course_title: string;
  toc: TocItem[];
  audience: string;
  template: TemplateKind;
  dos: string[];
  donts: string[];
}

export interface TocChange {
  action: string;
  target: string;
  proposed: string;
  reason: string;
}

export interface ImproveTocResponse {
  suggested_toc: TocItem[];
  changes: TocChange[];
  reasoning: string;
  missing_concepts: string[];
  duplicate_topics: string[];
}

// --- blueprint --------------------------------------------------------------

export interface ChapterSection {
  title: string;
  summary: string;
  key_points: string[];
}

export interface BlueprintChapter {
  id: string;
  title: string;
  order: number;
  objective: string;
  summary: string;
  sections: ChapterSection[];
  required_blocks: string[];
  optional_blocks: string[];
  key_concepts: string[];
  prerequisites: string[];
  research_questions: string[];
  estimated_words: number;
  difficulty: string;
}

export interface BlueprintCritique {
  missing_concepts: string[];
  duplicate_topics: string[];
  ordering_issues: string[];
  learning_progression_notes: string;
  suggested_toc: string[];
}

export interface CourseBlueprint {
  course_title: string;
  audience: string;
  template_id: string;
  course_summary: string;
  learning_objectives: string[];
  prerequisites: string[];
  tone: string;
  dos: string[];
  donts: string[];
  chapters: BlueprintChapter[];
  critique: BlueprintCritique;
  generated_at: string;
}

// --- generation -------------------------------------------------------------

export interface GenerateRequest {
  chapter_ids?: string[] | null;
  force?: boolean;
  skip_research?: boolean;
  generate_images?: boolean | null;
  build_document?: boolean;
  /** background returns a job id immediately; sync blocks until the run ends. */
  mode?: "background" | "sync";
  resume?: boolean;
  deep_research_chapter_ids?: string[] | null;
}

export interface GenerateResponse {
  course_id: string;
  document_id: string;
  status: CourseStatus;
  chapters_generated: string[];
  chapters_failed: string[];
  images_generated: number;
  pages: number;
  warnings: string[];
  duration_seconds: number;
  mode: "background" | "sync";
  job_id: string;
  accepted: boolean;
  timings: Record<string, unknown>;
}

export interface CourseDetail {
  course: CourseRecord;
  blueprint?: CourseBlueprint;
  artifacts: {
    blueprint: boolean;
    document: boolean;
    chapters: Array<{
      chapter_id: string;
      chapter_number: number;
      title: string;
      blocks: number;
      revisions: number;
      reviewed: boolean;
    }>;
    research: string[];
    pdf: string | null;
  };
}
